include/nbl/asset/IDescriptorSet.h

```diff
@@ -181,72 +187,27 @@ class NBL_API IEmulatedDescriptorSet
-				*it = {~0u,EDT_INVALID};
-			
-			auto outInfo = m_bindingInfo->begin();
-			uint32_t descriptorCount = 0u;
-			uint32_t prevBinding = 0;
-			// set up the offsets of specified bindings and determine descriptor count
-			for (auto it=bindings.begin(); it!=bindings.end(); it++)
+			for (uint32_t t = 0u; t < EDT_COUNT; ++t)
 			{
-				// if bindings are sorted, offsets shall be sorted too
-				assert(it==bindings.begin() || it->binding>prevBinding);
-
-				m_bindingInfo->operator[](it->binding) = { descriptorCount,it->type};
-				descriptorCount += it->count;
-				
-				prevBinding = it->binding;
-			}
+				const auto type = static_cast<E_DESCRIPTOR_TYPE>(t);
+				const uint32_t count = _layout->getTotalDescriptorCount(type);
+				if (count == 0u)
+					continue;
 
-			uint32_t offset = descriptorCount;
-			
-			m_descriptors = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<typename IDescriptorSet<LayoutType>::SDescriptorInfo> >(descriptorCount);
-			// set up all offsets, reverse iteration important because "it is for filling gaps with offset of next binding"
-			// TODO: rewrite this whole constructor to initialize the `SBindingOffset::offset` to 0 and simply use `std::exclusive_scan` to set it all up
-			for (auto it=m_bindingInfo->end()-1; it!=m_bindingInfo->begin()-1; it--)
-			{
-				if (it->offset < descriptorCount)
-					offset = it->offset;
-				else
-					it->offset = offset;
+				m_descriptors[type] = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<core::smart_refctd_ptr<IDescriptor>>>(count);
 			}
 
-			// this is vital for getDescriptorCountAtIndex
-			uint32_t off = ~0u;
-			for (auto it = m_bindingInfo->end() - 1; it != m_bindingInfo->begin() - 1; --it)
-			{
-				if (it->descriptorType != EDT_INVALID)
-					off = it->offset;
-				else
-					it->offset = off;
-			}
+			const uint32_t mutableSamplerCount = _layout->getTotalMutableSamplerCount();
+			if (mutableSamplerCount > 0u)
+				m_mutableSamplers = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<core::smart_refctd_ptr<ICPUSampler>>>(mutableSamplerCount);
 		}
 
 	protected:
 		virtual ~IEmulatedDescriptorSet() = default;
 
-		struct SBindingInfo
-		{
-			inline bool operator!=(const SBindingInfo& other) const
-			{
-				return offset!=other.offset || descriptorType!=other.descriptorType;
-			}
-
-			uint32_t offset;
-			E_DESCRIPTOR_TYPE descriptorType = EDT_INVALID;//whatever, default value
-		};
-		static_assert(sizeof(SBindingInfo)==8ull, "Why is the enum not uint32_t sized!?");
-		core::smart_refctd_dynamic_array<SBindingInfo> m_bindingInfo;
-		core::smart_refctd_dynamic_array<typename IDescriptorSet<LayoutType>::SDescriptorInfo> m_descriptors;
+	protected:
+		core::smart_refctd_dynamic_array<core::smart_refctd_ptr<IDescriptor>> m_descriptors[EDT_COUNT];
+		core::smart_refctd_dynamic_array<core::smart_refctd_ptr<ICPUSampler>> m_mutableSamplers = nullptr;
```

Comment:
I think, with the GL backend gone, we can move [this entire thing](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5ae62476fb0d73ac3ea4ce8e4a59ab21e8146483/include/nbl/asset/IDescriptorSet.h#L187-L210) to `ICPUDescriptorSet` and remove `IEmulatedDescriptorSet` entirely?

Reply:
no, how else are you going to have a WebGPU backend or something similar?

Also why is `m_mutableSamplers` an `ICPUSampler` should be an `typename layout_t::sampler_type`

Reply:
> Also why is `m_mutableSamplers` an `ICPUSampler` should be an `typename layout_t::sampler_type`

With `COpenGLDescriptorSet` gone, `ICPUDescriptorSet` was the only one that derived from `IEmulatedDescriptorSet` so I made the samplers `ICPUSampler`. But now thinking about a possible WebGPU backend implementation, I can see how this can be a problem. Using `typename layout_t::sampler_type` should better futureproof things.

Reply:
ah sorry, me dumb.

Move emulated descriptor set to `src/video` and slap a big fat comment that its unused and waiting for WebGPU

Reply:
again I'm dumb `ICPUDescriptorSet` needs the enumated one

again its best to move it out to a separate header, so its not sitting around in `IDescriptorSet`

Reply:
and remove the comment `//! Only reason this class exists is because OpenGL back-end implements a similar interface`

Reply:
> I think, with the GL backend gone, we can move [this entire thing](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5ae62476fb0d73ac3ea4ce8e4a59ab21e8146483/include/nbl/asset/IDescriptorSet.h#L187-L210) to `ICPUDescriptorSet` and remove `IEmulatedDescriptorSet` entirely?

Yes you are 100% correct.

I'm really dumb, sorry I've made you go around in circles.

You're right `IEmulatedDescriptorSet` has no right to exist really, `IDescriptorPool` covers everything for `IGPUDescriptorSet`

you can get rid of `IEmulatedDescriptorSet` and move all of its functionality to `ICPUDescriptorSet`

Reply:
I actually have a question, since we have `m_descriptorInfos[EDT_COUNT]` why would we even need `m_descriptors` from `IEmulatedDescriptorSet` in `ICPUDescriptorSet`?

Reply:
Right now, `m_descriptorInfos[i]` is a dynamic array of `SBufferImageInfo` (it doesn't contain the `core::smart_refctd_ptr<IDescriptor> desc` member).
If we make `m_descriptorInfos[i]` a dynamic array of `SDescriptorInfo` (which has the `core::smart_refctd_ptr<IDescriptor> desc` member) then I think we can get rid of the `m_descriptors` member.

I think the reason why we decided to do this is because of OpenGL. The idea could've been something like: store the descriptors (the common part) in `IEmulatedDescriptorSet` (common base for both `ICPUDescriptorSet` and `COpenGLDescriptorSet`) and the store their specific metadata in their respective derived classes, so `ICPUDescriptorSet` and `COpenGLDescriptorSet`. When we decided to deprecate the GL backend we just forgot to update this idea, that's all. No single person is to blame for it.

https://discord.com/channels/593902898015109131/888722959982813204/1031923212705091604 (Private Discord link)

Reply:
> If we make m_descriptorInfos[i] a dynamic array of SDescriptorInfo (which has the core::smart_refctd_ptr<IDescriptor> desc member) then I think we can get rid of the m_descriptors member.

yes thats best I think

Reply:
Resolved.